assert: add error diffs

BridgeAR opened the pull request:
The current assert errors are actually pretty bad when it comes to objects. This implements
a simple way to add diffs as default but only in assert strict mode. It relies on another PR that changes the `util.inspect` output as I use that for the diff. Building the diff differently would likely yield better results but it would also mean more work and this on its own should already be a significant improvement.

So far a few spots are not optimized for performance and I would like to get some general feedback on the solution. I especially do not like the way to distinguish `asserts` `strict mode` and `legacy mode`.
Activating the diff always would mean we have to change all our assert tests and I did not want to do that, besides the fact that some people might partially rely on the message output.

Refs #17576

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
assert, lib, util, errors

Trott commented on doc/api/assert.md:
```diff
@@ -26,12 +29,33 @@ When using the `strict mode`, any `assert` function will use the equality used i
 the strict function mode. So [`assert.deepEqual()`][] will, for example, work the
 same as [`assert.deepStrictEqual()`][].
 
+Besides that error messages that involve objects produce a error diff instead of
```
Nit: `a errror diff` -> `an error diff`

The sentence seems to be missing some punctuation for clarify. Probably a comma after the second word?

vsemozhetbyt replied:
Nit: a error -> an error

vsemozhetbyt commented on doc/api/assert.md:
````diff
@@ -26,12 +29,33 @@ When using the `strict mode`, any `assert` function will use the equality used i
 
+Example error diff:
+
+```diff
````
1. I am not sure, but maybe it is worth to separate the input with `js` label (for better highlighting and linting) and the output with `diff` label.

2. I cannot find any `` ```diff `` insets in `.md` files in this repo or `nodejs.org` repo. Are we sure they are rendered to the `.html` docs properly?

BridgeAR commented:
I just updated the code. I also improved the diffing algorithm from the first draft. It will now try to optimize the output rudimentary by removing identical lines from the back first and limits by output lines and some other smaller changes.

So PTAL

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12331/

BridgeAR commented:
Rebased due to conflicts.

I actually changed the tests to use the new `assert.throws(fn, obj)` pattern instead of `try {} catch {}` or `common.expectsError`.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/12348/

BridgeAR commented:
@nodejs/collaborators PTAL. This should be ready but it would be nice to get another LG.

mcollina commented on lib/internal/errors.js:
```diff
@@ -132,22 +132,164 @@ class SystemError extends makeNodeError(Error) {
+  const expectedLines = util
+    .inspect(expected, { compact: false, depth: 10 }).split('\n');
+  const msg = `Input A expected to ${operator} input B:\n` +
+        '\u001b[32m+ expected\u001b[39m \u001b[31m- actual\u001b[39m';
```
We should probably be testing that unix color codes are supported on the current terminal.

addaleax replied:
I don’t think we can test that but it might make sense to:
- Check whether `process.stderr` is a TTY
- Make coloring overridable in some way

BridgeAR replied:
That would indeed be nice... I will see what I can do but I am not sure if we can properly detect that.

addaleax commented on doc/api/assert.md:
```diff
@@ -26,12 +29,42 @@ When using the `strict mode`, any `assert` function will use the equality used i
 the strict function mode. So [`assert.deepEqual()`][] will, for example, work the
 same as [`assert.deepStrictEqual()`][].
 
+Besides that error messages that involve objects produce an error diff instead of
```
grammar nit: comma after the first `that`? Or maybe just start with `Error messages which involve objects …`?

addaleax commented on lib/assert.js:
```diff
@@ -152,7 +152,8 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
       operator: 'deepStrictEqual',
-      stackStartFn: deepStrictEqual
+      stackStartFn: deepStrictEqual,
+      errorDiff: this === strict ? 2 : 0
```
Can you name the different error diff modes? It’s really hard to keep track of constants `0`, `1` and `2` with distinct meanings, especially in code that spans across files.

BridgeAR commented:
@mcollina @addaleax I added a `util.hasColors` function. It is a very basic way to detect this but it seems like it is better than anything we had so far.

As this is probably also something that other users might want to rely on, I added it to the util module, even though I am not sure if it is the best place for it. What do you think about that?
And about adding it as a dedicated function in general: if we expose this, other users will report things to improve the current implementation :-).

@jasnell @Trott @vsemozhetbyt @TimothyGu please also take another look.

vsemozhetbyt commented on doc/api/util.md:
```diff
@@ -255,6 +255,28 @@ intended as a debugging tool. Some input values can have a significant
+added: REPLACEME
+-->
+
+* `stream` {Stream} A [stream][]. Defaults to `process.stdout`.
```
Missing bottom reference?

BridgeAR commented:
Rebased and new CI https://ci.nodejs.org/job/node-test-pull-request/12554/

I also addressed the doc comment.

BridgeAR commented:
Thanks for the reviews. This is going to be tedious though because lots of systems are failing because they all use different envs settings and support different things.

So I try to fix that now.

BridgeAR commented:
New CI for more debug infos: https://ci.nodejs.org/job/node-test-pull-request/12557/

BridgeAR commented:
I decided to remove the part that breaks a tests and to make that a separate PR.

New CI https://ci.nodejs.org/job/node-test-pull-request/12561/

jasnell commented on doc/api/assert.md:
````diff
@@ -26,12 +29,45 @@ When using the `strict mode`, any `assert` function will use the equality used i
+  ]
+```
+
+If you want to deactivate the colors, use the `NODE_DISABLE_COLORS` environment
````
nit: avoid using `you` in the docs

jasnell commented on doc/api/util.md:
```diff
@@ -257,6 +257,30 @@ intended as a debugging tool. Some input values can have a significant
+* 1 for 16,
+* 2 for 256,
+* 3 for 16,777,216
+colors supported.
```
Perhaps just make the return the actual values... e.g. return `0`, `16`, `256`, or `16777216`

BridgeAR replied:
Hm, I see your point but I personally think it is nicer to stick to the levels.

jasnell commented:
Couple of nits, but overall still lgtm

BridgeAR commented:
I addressed the comments and changed the return value to return the bit depths instead of levels. I also moved the function to the `tty` as it probably makes more sense there.

New CI https://ci.nodejs.org/job/node-test-pull-request/12589/

jasnell commented on doc/api/tty.md:
```diff
@@ -121,6 +121,29 @@ added: v0.7.7
+
+Use this to determine what colors the terminal supports.
+Due to the nature of colors in terminals it is possible to either have false
+positives or false negatives.
```
It might be worth noting in the docs that this determination is based on a set of inspection heuristics that may not be 100% accurate in every case. It really is dependent on the environment and the environment could, in fact, lie.

BridgeAR replied:
Done

jasnell commented on doc/api/tty.md:
```diff
@@ -121,6 +121,29 @@ added: v0.7.7
+positives or false negatives.
+
+Use the `NODE_DISABLE_COLORS` environment variable to enforce this function to
+always return 0.
```
Is this still true? The test seems to imply that it would return `1` if colors are disabled.

BridgeAR replied:
Solved.

jasnell commented on doc/api/tty.md:
```diff
@@ -137,12 +137,15 @@ Returns:
+colors in terminals it is possible to either have false positives or false
+negatives. It depends on process information and the environment variables that
+may lie about what terminal is used.
+In case you want to enforce a specific behavior you can pass in a object with
```
nit: avoid using `you` in the docs :-)

BridgeAR replied:
Arg... it always slips through :D

BridgeAR replied:
Done

BridgeAR commented:
One more CI https://ci.nodejs.org/job/node-test-pull-request/12612/

BridgeAR commented:
https://ci.nodejs.org/job/node-test-pull-request/12620/

apapirovski commented:
@BridgeAR Not sure if you saw on IRC but checkout `sequential/test-async-wrap-getasyncid.js` for the code to find the fd for TTY.

BridgeAR commented:
@apapirovski thanks a lot for the hint!

New CI: https://ci.nodejs.org/job/node-test-pull-request/12671/

BridgeAR commented:
This is now ready to land. @jasnell @apapirovski @mcollina @addaleax would you be so kind and take another look?

BridgeAR commented:
Landed in cb88f35a733c3d9efab5f7b18ca84047b93e7ab6...373e893ee0f09108289f623f73213907b8b05889

MylesBorins commented:
Should this be backported to `v9.x-staging`? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if not let me know or add the `dont-land-on` label.

BridgeAR commented:
Backport opened in #19230

sindresorhus commented on lib/tty.js:
```diff
@@ -92,39 +96,100 @@ function WriteStream(fd) {
+    return COLORS_16;
+
+  return COLORS_2;
+};
```
This code is mostly copy-pasted from https://github.com/chalk/supports-color/blob/master/index.js and https://github.com/isaacs/color-support/blob/master/index.js with minor style tweaks. I don't see any attribution.

richardlau replied:
@BridgeAR can you comment on the above?

BridgeAR replied:
Hey, I just came home and saw this, so I could not respond any earlier.

When writing on this I got the https://github.com/nodejs/node/pull/17615#discussion_r161598570 from @mcollina that I should check for the colors. So I googled a lot about checking color support in terminals and I tried to gather as much information as I could (I actually tried hard not to rely on environment variables at all). In the end I got inspired by multiple sources including `chalk`, `color-support`, different gists, stackoverflow and more that I do not remember. I did not think of anything bad when adding this function. I do see the similarities though and I am more than happy to give the attributions. I am just not sure how to do this properly right now as there are multiple licenses.

@sindresorhus shall I just move the function in a new file and add both license texts? I am not sure about the other sources though. Or would it be fine to add something like: `This functionality got inspired by multiple sources including `chalk` by @sindresorhus and `color-support` by @isaacs`?

BridgeAR replied:
Ping @isaacs

richardlau replied:
I guess also cc @nodejs/tsc FYI

joyeecheung replied:
@BridgeAR 
I think moving the color detection code to an internal lib file and put attributions there would be better than doing it in tty.js. But it still comes down to whether @sindresorhus and @isaacs accept it or not.

MylesBorins commented:
Requested backport to 8.x in #19230
